src/NServiceBus.RavenDB/Timeouts/TimeoutPersister.cs

```diff
@@ -43,43 +57,60 @@ public TimeoutPersister()
                 results = new List<Tuple<string, DateTime>>();
             }
 
-            // default return value for when no results are found
-            nextTimeToRunQuery = DateTime.UtcNow.AddMinutes(10);
+            RavenQueryStatistics statistics;
 
             using (var session = DocumentStore.OpenSession())
             {
-                var query = GetChunkQuery(session)
-                    .Where(t => t.Time > startSlice)
-                    .Select(t => new
-                    {
-                        t.Id,
-                        t.Time
-                    });
+                int totalCount, skipCount = 0;
 
-                QueryHeaderInformation qhi;
-                using (var enumerator = session.Advanced.Stream(query, out qhi))
+                do
                 {
-                    while (enumerator.MoveNext())
+                    var query = GetChunkQuery(session);
+
+                    var dueTimeouts =
+                        query.Statistics(out statistics)
+                            .Where(t => t.Time >= startSlice)
+                            .Where(t => t.Time < now)
+                            .Skip(skipCount)
+                            .Select(t => new
+                            {
+                                t.Id,
+                                t.Time
+                            })
+                            .Take(maximumPageSize);
+
+                    foreach (var dueTimeout in dueTimeouts)
```

Comment:
given that here we are materializing the query, and this can take time, is it worth to check for each item if there is a pending cancel request?